From c7b240d3516f34cbcc2abc17f7e0b98aa4d052df Mon Sep 17 00:00:00 2001
From: Mark Stapp <mjs@cisco.com>
Date: Wed, 29 Jul 2026 16:54:39 -0400
Subject: [PATCH 4/5] WIP adding resolving prefix in addition to the resolving
 NHG ID. Trying this out for now, adding cells to the nexthop struct; that may
 not be what we want to do permanently though.

Signed-off-by: Mark Stapp <mjs@cisco.com>
---
 lib/nexthop.c     | 14 +++++++++++++-
 lib/nexthop.h     |  4 +++-
 zebra/zebra_nhg.c | 46 ++++++++++++++++++++++++++++++++++++++++++----
 zebra/zebra_vty.c |  3 ++-
 4 files changed, 60 insertions(+), 7 deletions(-)

diff --git a/lib/nexthop.c b/lib/nexthop.c
index 2b49eaf31a..72113451c5 100644
--- a/lib/nexthop.c
+++ b/lib/nexthop.c
@@ -942,6 +942,9 @@ void nexthop_copy_no_recurse(struct nexthop *copy,
 
 	copy->srte_color = nexthop->srte_color;
 	copy->resolved_via = nexthop->resolved_via;
+	copy->resolved_addr = nexthop->resolved_addr;
+	copy->resolved_len = nexthop->resolved_len;
+
 	memcpy(&copy->gate, &nexthop->gate, sizeof(nexthop->gate));
 	memcpy(&copy->src, &nexthop->src, sizeof(nexthop->src));
 	memcpy(&copy->rmap_src, &nexthop->rmap_src, sizeof(nexthop->rmap_src));
@@ -1484,8 +1487,13 @@ void nexthop_json_helper(json_object *json_nexthop, const struct nexthop *nextho
 		}
 	}
 
-	if (nexthop->resolved_via > 0)
+	if (nexthop->resolved_via > 0) {
 		json_object_int_add(json_nexthop, "resolvedVia", nexthop->resolved_via);
+		json_object_string_addf(json_nexthop, "resolvedPrefix", "%pIA",
+					&(nexthop->resolved_addr));
+		json_object_int_add(json_nexthop, "resolvedPrefixLen",
+				    nexthop->resolved_len);
+	}
 }
 
 /*
@@ -1638,4 +1646,8 @@ void nexthop_vty_helper(struct vty *vty, const struct nexthop *nexthop,
 		for (i = 1; i < nexthop->backup_num; i++)
 			vty_out(vty, ",%d", nexthop->backup_idx[i]);
 	}
+
+	if (nexthop->resolved_via > 0)
+		vty_out(vty, ", res via %pIA/%d (%u)", &(nexthop->resolved_addr),
+			nexthop->resolved_len, nexthop->resolved_via);
 }
diff --git a/lib/nexthop.h b/lib/nexthop.h
index 5af44983e7..029bba6e7f 100644
--- a/lib/nexthop.h
+++ b/lib/nexthop.h
@@ -159,8 +159,10 @@ struct nexthop {
 	/* SR-TE color used for matching SR-TE policies */
 	uint32_t srte_color;
 
-	/* If resolved (by zebra typically), the NHG ID of the resolver */
+	/* If resolved (by zebra typically), the resolving NHG ID and prefix */
 	uint32_t resolved_via;
+	struct ipaddr resolved_addr;
+	uint8_t resolved_len;
 
 	/* SRv6 information */
 	struct nexthop_srv6 *nh_srv6;
diff --git a/zebra/zebra_nhg.c b/zebra/zebra_nhg.c
index 90e7958a05..893ef30178 100644
--- a/zebra/zebra_nhg.c
+++ b/zebra/zebra_nhg.c
@@ -2344,6 +2344,46 @@ zebra_nhg_connected_ifindex(struct route_node *rn, struct route_entry *match,
 	return match;
 }
 
+/*
+ * Locate NHG info for use as the "resolver" for a recursive route. This
+ * may mean navigating some of the NHG hierarchy; some levels of the NHG hierarchy
+ * are not installed through the dataplane, and aren't "visible" outside zebra.
+ */
+static void get_resolving_info(struct nexthop *nh, const struct route_node *rn,
+			       const struct route_entry *match)
+{
+	uint32_t id = match->nhe_id;
+	const struct nhg_hash_entry *nhe;
+	const struct nhg_connected *node;
+
+	nhe = match->nhe;
+
+	/* If the resolving nhg is itself recursive, it won't be installed. Walk
+	 * down the tree a bit to find the installed id value.
+	 */
+	if (CHECK_FLAG(nhe->flags, NEXTHOP_GROUP_RECURSIVE) &&
+	    !CHECK_FLAG(nhe->flags, NEXTHOP_GROUP_INSTALLED)) {
+		frr_each (nhg_connected_tree_const, &nhe->nhg_depends, node) {
+			if (CHECK_FLAG(node->nhe->flags, NEXTHOP_GROUP_INSTALLED)) {
+				id = node->nhe->id;
+				break;
+			}
+		}
+	}
+
+	nh->resolved_via = id;
+
+	/* Capture resolving prefix */
+	if (rn->p.family == AF_INET) {
+		SET_IPADDR_V4(&nh->resolved_addr);
+		nh->resolved_addr.ipaddr_v4 = rn->p.u.prefix4;
+	} else {
+		SET_IPADDR_V6(&nh->resolved_addr);
+		nh->resolved_addr.ipaddr_v6 = rn->p.u.prefix6;
+	}
+	nh->resolved_len = rn->p.prefixlen;
+}
+
 /*
  * Given a nexthop we need to properly recursively resolve,
  * do a table lookup to find and match if at all possible.
@@ -2674,11 +2714,9 @@ static int nexthop_active(struct nexthop *nexthop, struct nhg_hash_entry *nhe,
 				resolver = nexthop_set_resolved(afi, newhop, nexthop, NULL, flags);
 				resolved = 1;
 
-				/* Capture the NHG ID used to resolve the
-				 * nexthop.
-				 */
+				/* Capture the NHG ID used to resolve the nexthop. */
 				if (resolver)
-					resolver->resolved_via = match->nhe_id;
+					get_resolving_info(resolver, rn, match);
 
 				/* If there are backup nexthops, capture
 				 * that info with the resolving nexthop.
diff --git a/zebra/zebra_vty.c b/zebra/zebra_vty.c
index 674d800c16..d99f63dbc1 100644
--- a/zebra/zebra_vty.c
+++ b/zebra/zebra_vty.c
@@ -327,7 +327,8 @@ static void show_nexthop_detail_helper(struct vty *vty,
 	}
 
 	if (nexthop->resolved_via > 0)
-		vty_out(vty, ", res via %u", nexthop->resolved_via);
+		vty_out(vty, ", res via %pIA/%d (%u)", &(nexthop->resolved_addr),
+			nexthop->resolved_len, nexthop->resolved_via);
 }
 
 static void zebra_show_ip_route_opaque(struct vty *vty, struct route_entry *re,
-- 
2.50.1 (Apple Git-155)

